collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,87 @@
+import android.util.Log
+import com.bitmovin.analytics.data.AdEventData
+import com.bitmovin.analytics.data.Backend
+import com.bitmovin.analytics.data.CallbackBackend
+import com.bitmovin.analytics.data.EventData
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+import java.util.Date
+import kotlin.Exception
+import okhttp3.internal.http2.StreamResetException
+
+class RetryBackend(private val next: CallbackBackend, val handler: Handler) : Backend {
+
+    private val TAG = "RetryBackend"
+    private var retryDateToken: Date? = null
+    private val retryQueue = RetryQueue()
+    fun getNextScheduledTime() = retryQueue.getNextScheduleTime()
+
+    override fun send(eventData: EventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    override fun sendAd(eventData: AdEventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    private fun scheduleSample(retrySample: RetrySample<Any>) {
+        val callback = object : OnFailureCallback {
+            override fun onFailure(e: Exception, cancel: () -> Unit) {
+                if (e is SocketTimeoutException || e is ConnectException || e is StreamResetException || e is UnknownHostException) {
+                    cancel()
+                    retryQueue.addSample(retrySample)
+                    processQueuedSamples()
+                }
+            }
+        }
+
+        if (retrySample.eventData is EventData) {
+            Log.d(TAG, "sending sample ${retrySample.eventData?.sequenceNumber} retry ${retrySample.retry}")
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.send(retrySample.eventData, callback)
+        } else if (retrySample.eventData is AdEventData) {
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.sendAd(retrySample.eventData, callback)
+        }
+    }
+
+    @Synchronized
+    fun processQueuedSamples() {
+        try {
+            val nextScheduledTime = getNextScheduledTime()
+            if (nextScheduledTime != null || retryDateToken != nextScheduledTime) {
+                if (retryDateToken != null) {
+                    handler.removeCallbacks(processSampleRunnable, retryDateToken)
+                }
+                retryDateToken = nextScheduledTime
+                val delay = maxOf((nextScheduledTime?.time ?: 0) - Date().time, 0) // to prevent negative delay
+                handler.postAtTime(processSampleRunnable, retryDateToken, SystemClock.uptimeMillis() + delay)
+            }
+        } catch (e: Exception) {
+            Log.d(TAG, "processQueuedSamples ${e.message}")
+        }
+    }
+
+    private val processSampleRunnable = Runnable {
+        try {
+            val retrySample = retryQueue.getNextSampleOrNull()
+            if (retrySample != null) {
+                scheduleSample(retrySample)
+            }
+            retryDateToken = null
```

Comment:
What is the reasoning for naming the field `token`?

Reply:
Logic for re-sending samples uses handlers for scheduling execution of logic for getting and sending next sample.
https://developer.android.com/reference/android/os/Handler#postAtTime(java.lang.Runnable,%20java.lang.Object,%20long)

Method used `postAtTime` receives 3 parameters:
- runnable (interface that wraps code which should be executed)
- token (an instance which can be used to cancel runnable, can be any object)
- uptimeMillis (when runnable should be executed)

So I wanted to indicate in variable name as what this variable is used. I can also rename it to something like `nextScheduledExecutionTime`?

Reply:
But this is not a time then but rather a `cancelToken`? Or are you just using the same object for storing the executionTime + the cancel Token ?

Reply:
Well both... I save next scheduled time in this variable so I can check if there is some sample which should be sent before, if there is then I use it as cancel token.